### DISCUSSION TITLE
Modernize Aardvark?

### DISCUSSION THREAD

**Discussion by cocoalabs** [square/Aardvark#139] [General]
Currently Aardvark supports iOS 12. I'm proposing that we move the minimum deployment target to iOS 14 and modernize it by adding the following:

- Port entirely to Swift
- Adopt async/await
-  Adopt structured concurrency
- Support more flexible encoding options for attachments (for example: support zipping of any Data encodable object).
- Support additional transports other than email. i.e. HTTP post,  messaging queues such as Kafka/Rabbit MQ etc. (general extensible architecture for posting resultant bug reports)

**Comment by dfed** [comment]
I love the idea of modernizing Aardvark. Since Aardvark has been stable for quite some time, rolling a major version doesn't have much in the way of downsides and will make maintaining Aardvark easier long-term.

If I can save you a little bit of time, consider depending on the following repos in the new iteration of Aardvark:
* [CacheAdvance](https://github.com/dfed/cacheadvance). This repo vends a type that was inspired by `ARKDataArchive`, but is significantly faster, trims older messages less aggressively while still maintaining a maximum size, and is written in Swift.
* [AsyncQueue](https://github.com/dfed/swift-async-queue). This repo enables transitioning from nonisolated to isolated contexts in Swift in an ordered manner. You'll have trouble writing an ordered logging system that utilizes Swift Concurrency without it or something similar.

Some food for thought as you embark on this adventure: I always felt that one of my original sins when creating Aardvark was that the project was too much of a monolith. The repo is responsible for doing too many things under one umbrella: Aardvark is a great incremental log store, a great logging system with log filtering, and a great bug filing tool. Given the level of customizability desired by different consumers, I've felt for awhile that creating separate repos – one for each core competency – would have helped enforce separation of concerns (which we had to build anyways!) and better unlocked contributions from the greater community.

I'm really excited that y'all have made Aardvark your own. Hopefully the above is helpful perspective from the guy who originally poured his soul into this repo, but don't let my thoughts hold you back: this repo is your baby now!
